From 918d200146c15c6f888789981cc511ae2983fc80 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 19 Jan 2021 11:16:36 +0800
Subject: [PATCH] waylandsink: Defer prepare window when getting zero window
 handle

The window might not ready when we requesting it.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 ext/wayland/gstwaylandsink.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 78f57c1..1c08466 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -703,9 +703,16 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
     /* ask for window handle. Unlock render_lock while doing that because
      * set_window_handle & friends will lock it in this context */
     g_mutex_unlock (&sink->render_lock);
+    sink->window_handle = 1;
     gst_video_overlay_prepare_window_handle (GST_VIDEO_OVERLAY (sink));
     g_mutex_lock (&sink->render_lock);
 
+    /* HACK: Defer window prepare when getting zero window handle */
+    if (!sink->window_handle) {
+      GST_LOG_OBJECT (sink, "buffer %p dropped (window not ready)", buffer);
+      goto done;
+    }
+
     if (!sink->window) {
       /* if we were not provided a window, create one ourselves */
       sink->window = gst_wl_window_new_toplevel (sink->display,
-- 
2.20.1

